syz-manager/repro.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"context"
+	"maps"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/instance"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/stats"
+	"github.com/google/syzkaller/vm"
+)
+
+type repoManagerView interface {
+	// TODO: consider moving the function to reproManager.
+	runRepro(crash *Crash) *ReproResult
+}
+
+type reproManager struct {
+	Pending            atomic.Int64
+	StatNumReproducing *stats.Val
+	Done               chan *ReproResult
+
+	pool *instance.Pool[*vm.Instance]
+
+	onlyOnce bool
+
+	mu          sync.Mutex
+	ctx         context.Context
+	mgr         repoManagerView
+	reproducing map[string]bool
+	attempted   map[string]bool
+	parallel    chan struct{}
+}
+
+func newReproManager(ctx context.Context, pool *instance.Pool[*vm.Instance],
+	onlyOnce bool, mgr repoManagerView) *reproManager {
+	ret := &reproManager{
+		StatNumReproducing: stats.Create("reproducing", "Number of crashes being reproduced",
+			stats.Console, stats.NoGraph),
+		Done: make(chan *ReproResult, 10),
+
+		ctx:         ctx,
+		pool:        pool,
+		onlyOnce:    onlyOnce,
+		mgr:         mgr,
+		parallel:    make(chan struct{}, pool.Total()),
+		reproducing: map[string]bool{},
+		attempted:   map[string]bool{},
+	}
+	go ret.adjustPoolSize()
+	return ret
+}
+
+// startReproduction() is assumed to be called only once.
+// The agument is the maximum number of VMs dedicated to the bug reproduction.
+func (m *reproManager) startReproduction(VMs int) {
+	log.Logf(1, "starting reproductions (%d VMs available)", VMs)
+	for count := 1; count <= VMs; count++ {
+		need := m.needVMs(count)
+		if need > VMs {
+			break
+		}
+		m.parallel <- struct{}{}
+	}
+}
+
+func (m *reproManager) needVMs(repros int) int {
+	// For now, let's allocate 1.33 VMs per a reproducer thread.
+	return (repros*4 + 2) / 3
+}
+
+func (m *reproManager) canReproMore() bool {
+	select {
+	case <-m.parallel:
+		m.parallel <- struct{}{}
+		return true
+	default:
+		return false
+	}
+}
+
+func (m *reproManager) reproducingNow() map[string]bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return maps.Clone(m.reproducing)
+}
+
+func (m *reproManager) handle(crash *Crash) {
+	m.mu.Lock()
+	// We don't want to reproduce the same bug twice at the same time.
+	if m.reproducing[crash.Title] {
+		m.mu.Unlock()
+		return
```

Comment:
I am not sure this is correct. If there are 2 crashes for the same bug, we don't want to run repro in parallel. But if the first has failed, we still want to run the second one later.
Currently there is a queue in manager.go for this case.
But now we discard the second crash forever, if first repro is running.